### PR TITLE
Add Segment Adapter

### PR DESCRIPTION
**Pull request by seanpdoyle**
Implements a [Segment.js] wrapper. In terms of implementation, the
`ember-metrics` adapter API maps nicely to the Segment.js API:

* Ember-Metric's `identify` -> Segment's `identify`
* Ember-Metric's `alias` -> Segment's `alias`
* Ember-Metric's `trackEvent` -> Segment's `track`
* Ember-Metric's `trackPage` -> Segment's `page`

[Segment.js]: https://segment.com/docs/libraries/analytics.js/

### CODE REVIEW

File: addon/metrics-adapters/segment.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+  identify(options = {}) {
+    const compactedOptions = compact(options);
+    const { distinctId } = compactedOptions;
+    delete compactedOptions.distinctId;
+
+    window.analytics.identify(distinctId, compactedOptions);
+  },
+
+  trackEvent(options = {}) {
```
Comment (croaky) [inline addon/metrics-adapters/segment.js]: Should this be `track` to be consistent with the analytics.js API?

File: addon/metrics-adapters/segment.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    const compactedOptions = compact(options);
+    const { event } = compactedOptions;
+    delete compactedOptions.event;
+
+    window.analytics.track(event, compactedOptions);
+  },
+
+  trackPage(options = {}) {
```
Comment (croaky) [inline addon/metrics-adapters/segment.js]: Should this be `page` to be consistent with the analytics.js API?

File: addon/metrics-adapters/segment.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    const { page } = compactedOptions;
+    delete compactedOptions.page;
+
+    window.analytics.page(page, compactedOptions);
+  },
+
+  willDestroy() {
+    $('script[src*="segment.com"]').remove();
+    delete window.analytics;
+  }
```
Comment (croaky) [inline addon/metrics-adapters/segment.js]: The two API methods I use regularly that aren't here yet are `trackLink` and `trackForm`. Could be their own PRs, not sure how TODOs are captured on this project.

**Comment by seanpdoyle** [comment]
@croaky sorry, I should've given you more context on what this PR represents.

If you checkout the [README](https://github.com/poteto/ember-metrics#service-api), these methods are defined on the metrics service. 

They're wrappers around the adapter that you're using. For instance, the metrics service responds to `trackPage`, regardless of the backing implementation. This means that `service.trackPage({ page: 'Sign Up' })` works the same for Segment as it does for Mixpanel as it does for GoogleAnalytics.

File: addon/metrics-adapters/segment.js (hunk 3 above)
Comment (seanpdoyle) [inline addon/metrics-adapters/segment.js]: I think since link navigation and form submission happen client side (via Ember's actions and Router transitions), `trackPage` and `trackEvent` should cover their purposes respectively.

Would you agree?

File: README.md [hunk 4]
```diff
@@ -25,8 +25,10 @@ Writing your own adapters for currently unsupported analytics services is easy t
       'visitorType': 'high-value'
     }];
     ```
-4. `KISSMetrics` (WIP)
-5. `CrazyEgg` (WIP)
+1. `Segment`
+  - `key`: [Segment key](https://segment.com/docs/libraries/analytics.js/quickstart/)
+1. `KISSMetrics` (WIP)
+1. `CrazyEgg` (WIP)
```
Comment (croaky) [inline README.md]: Are you thinking of not directly supporting these other services, pushing people to go through Segment to get this functionality? Less to maintain in the repo.

**Comment by croaky** [comment]
@seanpdoyle Cool.

Code looks good to me.

The times I've seen Segment added to Ember, the main gotcha was that the `page` event wasn't being triggered in the right spot. I don't know much about Ember, but I would guess it has to be tied into the/a `Router` or something?

**Comment by seanpdoyle** [comment]
@croaky yup, the [README](https://github.com/poteto/ember-metrics#usage) walks you through how to extend `Router#didTransition` to fire the proper calls.

File: addon/metrics-adapters/segment.js (hunk 3 above)
Comment (poteto) [inline addon/metrics-adapters/segment.js]: `trackLink` and `trackForm` are sugar for adding delays to click and submit events. In Ember, you can simply use the `trackEvent` method on the metrics service when handling actions.

File: addon/metrics-adapters/segment.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    return 'Segment';
+  },
+
+  init() {
+    const config = copy(get(this, 'config'));
+    const segmentKey = config.key;
+
+    assert(`[ember-metrics] You must pass a valid \`key\` to the ${this.toString()} adapter`, segmentKey);
+
+    delete config.key;
```
Comment (poteto) [inline addon/metrics-adapters/segment.js]: This doesn't seem necessary

File: addon/metrics-adapters/segment.js [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+    if (canUseDOM) {
+      /* jshint ignore:start */
+      window.analytics=window.analytics||[],window.analytics.methods=["identify","group","track","page","pageview","alias","ready","on","once","off","trackLink","trackForm","trackClick","trackSubmit"],window.analytics.factory=function(t){return function(){var a=Array.prototype.slice.call(arguments);return a.unshift(t),window.analytics.push(a),window.analytics}};for(var i=0;i<window.analytics.methods.length;i++){var key=window.analytics.methods[i];window.analytics[key]=window.analytics.factory(key)}window.analytics.load=function(t){if(!document.getElementById("analytics-js")){var a=document.createElement("script");a.type="text/javascript",a.id="analytics-js",a.async=!0,a.src=("https:"===document.location.protocol?"https://":"http://")+"cdn.segment.com/analytics.js/v1/"+t+"/analytics.min.js";var n=document.getElementsByTagName("script")[0];n.parentNode.insertBefore(a,n)}},window.analytics.SNIPPET_VERSION="2.0.9";
+      /* jshint ignore:end */
+      window.analytics.load(segmentKey);
+      window.analytics.page();
```
Comment (poteto) [inline addon/metrics-adapters/segment.js]: Why is this tracking the page view on init?

**Comment by poteto** [comment]
Thanks for this PR!

File: addon/metrics-adapters/segment.js (hunk 3 above)
Comment (seanpdoyle) [inline addon/metrics-adapters/segment.js]: :clap:

File: addon/metrics-adapters/segment.js (hunk 5 above)
Comment (seanpdoyle) [inline addon/metrics-adapters/segment.js]: Stolen [directly from here](https://github.com/poteto/ember-metrics/blob/11a1bb61a350e74cc5d8c585bb595aa6637ad42d/addon/metrics-adapters/google-analytics.js#L28)

Comment (poteto) [inline addon/metrics-adapters/segment.js]: Ah, yes. `config` [was being passed into `window.ga`](https://github.com/poteto/ember-metrics/blob/11a1bb61a350e74cc5d8c585bb595aa6637ad42d/addon/metrics-adapters/google-analytics.js#L41), so removing the `id` prop on the object was necessary to prevent errors being thrown by analytics.js. In this adapter, it does not appear that the Segment library takes options when being initialized, so deleting it seems unnecessary

**Comment by poteto** [comment]
LGTM!
